AzureIoTHub.agent.lib.nut

```diff
@@ -629,513 +621,834 @@ class AzureIoTHub {
-                                    _openEventSender();
-                                }
-                            } else {
-                                _openEventSession();
-                            }
-                        } else {
-                            _openAuthSender();
-                        }
-                    } else {
-                        _openAuthReceiver();
-                    }
-                } else {
-                    _openAuthSession();
-                }
-            } else {
-                _connection = amqp.openconnection("amqps://" + _config.host, _amqpConnectionStatusHandler.bindenv(this));
+            _onConnectCb      = onConnect;
+            _onDisconnectCb   = onDisconnect;
+
+            _connStrParsed = AzureIoTHub.ConnectionString.Parse(deviceConnStr);
+            _mqttclient = mqtt.createclient();
+            _mqttclient.onconnect(_onConnect.bindenv(this));
+            _mqttclient.onconnectionlost(_onDisconnect.bindenv(this));
+            _mqttclient.onmessage(_onMessage.bindenv(this));
+
+            foreach (optName, optVal in options) {
+                _options[optName] <- optVal;
             }
-        }
 
-        function disconnect() {
-            _connection.close();
-            _resetConnectionTables();
+            _fillTopics(_connStrParsed.DeviceId);
         }
 
-        // done cb params - err
-        function sendEvent(message, done = null) {
-            if ( !_isOpen(_sessions.event) ) {
-                if (done) done("Cannot send while disconnected.");
-            } else {
-                if ( !_isOpen(_senders.event) ) {
-                    _log("creating event sender");
-                    // add message to queue
-                    _msgQueue.push({"msg" : message, "cb" : done});
-                    // create a new sender & send message
-                    _openEventSender();
-                } else {
-                    _log("event sender open")
-                    _sendEvent(message, done);
-                }
+        // Opens a connection to Azure IoT Hub.
+        //
+        // Returns:                         Nothing.
+        function connect() {
+            if (_isConnected || _isConnecting) {
+                _onConnectCb(_isConnected ? AZURE_IOT_CLIENT_ERROR_ALREADY_CONNECTED : AZURE_IOT_CLIENT_ERROR_OP_NOT_ALLOWED_NOW);
+                return;
             }
-        }
+            _log("Connecting...");
+
+            local devPath = "/" + _connStrParsed.DeviceId;
+            local username = format("%s%s/api-version=%s", _connStrParsed.HostName, devPath, AZURE_API_VERSION);
+            local resourcePath = format("/devices%s/api-version=%s", devPath, AZURE_API_VERSION);
+            local resourceUri = AzureIoTHub.Authorization.encodeUri(_connStrParsed.HostName + resourcePath);
+            local passwDeadTime = AzureIoTHub.Authorization.anHourFromNow();
+            local sas = AzureIoTHub.SharedAccessSignature.create(
```

Comment:
(cosmetic, not sure if that was you who wrote that part initially) I would probably get rid of the factory `AzureIoTHub.SharedAccessSignature.create` method and make it constructor instead.

Reply:
Done